File: app/src/main/java/com/jobtick/fragments/ProfileReqFragment.java
package com.jobtick.fragments;

import android.content.Intent;
import android.os.Bundle;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.google.android.material.button.MaterialButton;
import com.jobtick.R;

import com.jobtick.activities.AbstractUploadableImageImpl;
import com.jobtick.activities.ActivityBase;
import com.jobtick.activities.TaskDetailsActivity;
import com.jobtick.activities.UploadableImage;
import com.jobtick.adapers.AttachmentAdapter;
import com.jobtick.models.AttachmentModel;
import com.jobtick.models.UserAccountModel;
import com.jobtick.retrofit.ApiClient;
import com.jobtick.utils.HttpStatus;
import com.jobtick.utils.ImageUtil;
import com.jobtick.utils.SessionManager;
import com.mikhaellopez.circularimageview.CircularImageView;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;

import butterknife.ButterKnife;
import butterknife.OnClick;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Callback;
import timber.log.Timber;

import static com.jobtick.activities.DashboardActivity.onProfileupdatelistenerSideMenu;
import static com.jobtick.fragments.ProfileFragment.onProfileupdatelistener;
import static com.jobtick.utils.ConstantKey.TAG;

public class ProfileReqFragment extends Fragment {

    private CircularImageView imgAvatar;
    SessionManager sessionManager;
    private UserAccountModel userAccountModel;
    MaterialButton btnNext;

    private UploadableImage uploadableImage;

    public ProfileReqFragment() {
    }

    public static ProfileReqFragment newInstance() {
        return new ProfileReqFragment();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        sessionManager = new SessionManager(getContext());
        imgAvatar = view.findViewById(R.id.img_user_avatar);
        btnNext = view.findViewById(R.id.btn_next);
        btnNext.setOnClickListener(v -> {
            ((TickerRequirementsBottomSheet) getParentFragment()).changeFragment(1);
        });

        userAccountModel = ((TaskDetailsActivity) getActivity()).userAccountModel;
        setUpAvatar(userAccountModel);

        uploadableImage = new AbstractUploadableImageImpl(requireActivity()) {
            @Override
            public void onImageReady(File imageFile) {
                uploadProfileAvatar(imageFile);
            }
        };
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_profile_req, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    private void setUpAvatar(UserAccountModel userAccountModel) {
        try {
            ImageUtil.displayImage(imgAvatar, userAccountModel.getAvatar().getUrl(), null);
        } catch (Exception ex) {
            Timber.tag(TAG).e("EXCEPTION CAUGHT WHILE EXECUTING DATABASE TRANSACTION");
            ex.printStackTrace();
        }
    }

    @OnClick({R.id.img_user_avatar})
    public void onViewClicked(View view) {
        if (view.getId() == R.id.img_user_avatar) {
            uploadableImage.showAttachmentBottomSheet(true);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        uploadableImage.onActivityResult(requestCode, resultCode, data);
    }

    private void uploadProfileAvatar(File pictureFile) {
        Call<String> call;
        RequestBody requestFile = RequestBody.create(MediaType.parse("multipart/form-data"), pictureFile);
        MultipartBody.Part imageFile = MultipartBody.Part.createFormData("media", pictureFile.getName(), requestFile);
        call = ApiClient.getClient().uploadProfilePicture("XMLHttpRequest", sessionManager.getTokenType() + " " + sessionManager.getAccessToken(), imageFile);
        call.enqueue(new Callback<String>() {
            @Override
            public void onResponse(Call<String> call, retrofit2.Response<String> response) {
                String res = response.body();
                System.out.println(res);
                if (response.code() == HttpStatus.HTTP_VALIDATION_ERROR) {
                    if (getActivity() != null) {
                        ((ActivityBase) getActivity()).showToast(response.message(), getActivity());
                    }
                    return;
                }
                try {
                    String strResponse = res;
                    if (response.code() == HttpStatus.NOT_FOUND) {
                        ((ActivityBase)requireActivity()).showToast("not found", requireContext());
                        return;
                    }
                    if (response.code() == HttpStatus.AUTH_FAILED) {
                        ((ActivityBase) getActivity()).unauthorizedUser();
                        return;
                    }
                    if (response.code() == HttpStatus.SUCCESS) {
                        JSONObject jsonObject = new JSONObject(strResponse);
                        if (jsonObject.has("data")) {
                            AttachmentModel attachment = new AttachmentModel();
                            JSONObject jsonObject_data = jsonObject.getJSONObject("data");
                            if (jsonObject_data.has("id") && !jsonObject_data.isNull("id"))
                                attachment.setId(jsonObject_data.getInt("id"));
                            if (jsonObject_data.has("name") && !jsonObject_data.isNull("name"))
                                attachment.setName(jsonObject_data.getString("name"));
                            if (jsonObject_data.has("file_name") && !jsonObject_data.isNull("file_name"))
                                attachment.setFileName(jsonObject_data.getString("file_name"));
                            if (jsonObject_data.has("mime") && !jsonObject_data.isNull("mime"))
                                attachment.setMime(jsonObject_data.getString("mime"));
                            if (jsonObject_data.has("url") && !jsonObject_data.isNull("url"))
                                attachment.setUrl(jsonObject_data.getString("url"));
                            if (jsonObject_data.has("thumb_url") && !jsonObject_data.isNull("thumb_url"))
                                attachment.setThumbUrl(jsonObject_data.getString("thumb_url"));
                            if (jsonObject_data.has("modal_url") && !jsonObject_data.isNull("modal_url"))
                                attachment.setModalUrl(jsonObject_data.getString("modal_url"));
                            if (jsonObject_data.has("created_at") && !jsonObject_data.isNull("created_at"))
                                attachment.setCreatedAt(jsonObject_data.getString("created_at"));
                            attachment.setType(AttachmentAdapter.VIEW_TYPE_IMAGE);

                            sessionManager.getUserAccount().setAvatar(attachment);
                            ImageUtil.displayImage(imgAvatar, attachment.getUrl(), null);

                            if (onProfileupdatelistener != null) {
                                onProfileupdatelistener.updatedSuccesfully(attachment.getUrl());
                            }
                            if (onProfileupdatelistenerSideMenu != null) {
                                onProfileupdatelistenerSideMenu.updatedSuccesfully(attachment.getUrl());
                            }
                        }
                        //   adapter.notifyItemRangeInserted(0,attachmentArrayList.size());
                        // ((ActivityBase) getActivity()).showToast("attachment added", getActivity());
                        btnNext.setEnabled(true);
                    } else {
                        ((ActivityBase) getActivity()).showToast("Something went wrong", getActivity());
                    }
                } catch (JSONException e) {
                    ((ActivityBase) getActivity()).showToast("Something went wrong", getActivity());
                    e.printStackTrace();
                }
            }

            @Override
            public void onFailure(Call<String> call, Throwable t) {
                if (getActivity() != null) {
                    ((ActivityBase) getActivity()).hideProgressDialog();
                }
            }
        });
    }
}

File: app/src/main/java/com/jobtick/android/fragments/ProfileReqFragment.kt
package com.jobtick.android.fragments

import com.mikhaellopez.circularimageview.CircularImageView
import com.jobtick.android.models.UserAccountModel
import com.google.android.material.button.MaterialButton
import com.jobtick.android.activities.UploadableImage
import android.os.Bundle
import com.jobtick.android.R
import com.jobtick.android.activities.TaskDetailsActivity
import com.jobtick.android.activities.AbstractUploadableImageImpl
import android.view.LayoutInflater
import android.view.ViewGroup
import butterknife.ButterKnife
import timber.log.Timber
import com.jobtick.android.utils.ConstantKey
import butterknife.OnClick
import android.content.Intent
import android.view.View
import androidx.fragment.app.Fragment
import okhttp3.RequestBody
import okhttp3.MultipartBody
import com.jobtick.android.retrofit.ApiClient
import com.jobtick.android.utils.HttpStatus
import com.jobtick.android.activities.ActivityBase
import com.jobtick.android.models.AttachmentModel
import com.jobtick.android.adapers.AttachmentAdapter
import com.jobtick.android.activities.DashboardActivity
import org.json.JSONException
import com.jobtick.android.utils.ImageUtil
import com.jobtick.android.utils.SessionManager
import okhttp3.MediaType
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.File
import java.lang.Exception

class ProfileReqFragment : Fragment() {
    private var imgAvatar: CircularImageView? = null
    var sessionManager: SessionManager? = null
    private var userAccountModel: UserAccountModel? = null
    var btnNext: MaterialButton? = null
    private var uploadableImage: UploadableImage? = null
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        sessionManager = SessionManager(context)
        imgAvatar = view.findViewById(R.id.img_user_avatar)
        imgAvatar!!.setOnClickListener {
            if (view.id == R.id.img_user_avatar) {
                uploadableImage!!.showAttachmentImageBottomSheet(true)
            }
        }
        btnNext = view.findViewById(R.id.btn_next)
        btnNext!!.setOnClickListener(View.OnClickListener { v: View? -> (parentFragment as TickerRequirementsBottomSheet?)!!.changeFragment(1) })
        userAccountModel = (activity as TaskDetailsActivity?)!!.userAccountModel
        setUpAvatar(userAccountModel)
        uploadableImage = object : AbstractUploadableImageImpl(requireActivity()) {
            override fun onImageReady(imageFile: File) {
                uploadProfileAvatar(imageFile)
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_profile_req, container, false)
        ButterKnife.bind(this, view)
        return view
    }

    private fun setUpAvatar(userAccountModel: UserAccountModel?) {
        try {
            ImageUtil.displayImage(imgAvatar, userAccountModel!!.avatar.url, null)
        } catch (ex: Exception) {
            Timber.tag(ConstantKey.TAG).e("EXCEPTION CAUGHT WHILE EXECUTING DATABASE TRANSACTION")
            ex.printStackTrace()
        }
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        uploadableImage!!.onActivityResult(requestCode, resultCode, data)
    }

    private fun uploadProfileAvatar(pictureFile: File) {
        val call: Call<String?>?
        val requestFile = RequestBody.create(MediaType.parse("multipart/form-data"), pictureFile)
        val imageFile = MultipartBody.Part.createFormData("media", pictureFile.name, requestFile)
        call = ApiClient.getClient().uploadProfilePicture("XMLHttpRequest", sessionManager!!.tokenType + " " + sessionManager!!.accessToken, imageFile)
        call!!.enqueue(object : Callback<String?> {
            override fun onResponse(call: Call<String?>, response: Response<String?>) {
                val res = response.body()
                println(res)
                if (response.code() == HttpStatus.HTTP_VALIDATION_ERROR) {
                    if (activity != null) {
                        (activity as ActivityBase?)!!.showToast(response.message(), activity)
                    }
                    return
                }
                try {
                    if (response.code() == HttpStatus.NOT_FOUND) {
                        (requireActivity() as ActivityBase).showToast("not found", requireContext())
                        return
                    }
                    if (response.code() == HttpStatus.AUTH_FAILED) {
                        (activity as ActivityBase?)!!.unauthorizedUser()
                        return
                    }
                    if (response.code() == HttpStatus.SUCCESS) {
                        val jsonObject = JSONObject(res)
                        if (jsonObject.has("data")) {
                            val attachment = AttachmentModel()
                            val jsonObjectData = jsonObject.getJSONObject("data")
                            if (jsonObjectData.has("id") && !jsonObjectData.isNull("id")) attachment.id = jsonObjectData.getInt("id")
                            if (jsonObjectData.has("name") && !jsonObjectData.isNull("name")) attachment.name = jsonObjectData.getString("name")
                            if (jsonObjectData.has("file_name") && !jsonObjectData.isNull("file_name")) attachment.fileName = jsonObjectData.getString("file_name")
                            if (jsonObjectData.has("mime") && !jsonObjectData.isNull("mime")) attachment.mime = jsonObjectData.getString("mime")
                            if (jsonObjectData.has("url") && !jsonObjectData.isNull("url")) attachment.url = jsonObjectData.getString("url")
                            if (jsonObjectData.has("thumb_url") && !jsonObjectData.isNull("thumb_url")) attachment.thumbUrl = jsonObjectData.getString("thumb_url")
                            if (jsonObjectData.has("modal_url") && !jsonObjectData.isNull("modal_url")) attachment.modalUrl = jsonObjectData.getString("modal_url")
                            if (jsonObjectData.has("created_at") && !jsonObjectData.isNull("created_at")) attachment.createdAt = jsonObjectData.getString("created_at")
                            attachment.type = AttachmentAdapter.VIEW_TYPE_IMAGE
                            sessionManager!!.userAccount.avatar = attachment
                            ImageUtil.displayImage(imgAvatar, attachment.url, null)
                            if (ProfileFragment.onProfileupdatelistener != null) {
                                ProfileFragment.onProfileupdatelistener!!.updatedSuccesfully(attachment.url)
                            }
                            if (DashboardActivity.onProfileupdatelistenerSideMenu != null) {
                                DashboardActivity.onProfileupdatelistenerSideMenu!!.updatedSuccesfully(attachment.url)
                            }
                        }
                        //   adapter.notifyItemRangeInserted(0,attachmentArrayList.size());
                        // ((ActivityBase) getActivity()).showToast("attachment added", getActivity());
                        btnNext!!.isEnabled = true
                    } else {
                        (activity as ActivityBase?)!!.showToast("Something went wrong", activity)
                    }
                } catch (e: JSONException) {
                    (activity as ActivityBase?)!!.showToast("Something went wrong", activity)
                    e.printStackTrace()
                }
            }

            override fun onFailure(call: Call<String?>, t: Throwable) {
                if (activity != null) {
                    (activity as ActivityBase?)!!.hideProgressDialog()
                }
            }
        })
    }

    companion object {
        @JvmStatic
        fun newInstance(): ProfileReqFragment {
            return ProfileReqFragment()
        }
    }
}